File: Binary Subarrays With Sum.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        if goal == 0:
            combo = 0
            total = 0
            for num in nums:
                if num == 1 and combo > 0:
                    total += (combo * (combo + 1)) // 2
                    combo = 0
                elif num == 0:
                    combo += 1

            if combo > 0:
                total += (combo * (combo + 1)) // 2
            return total

        arr = []
        combo = 1
        for num in nums:
            if num == 1:
                arr.append(combo)
                combo = 1
            else:
                combo += 1

        arr.append(combo)
        left, right = 0, goal
        total = 0
        while right < len(arr):
            total += arr[right] * arr[left]
            right += 1
            left += 1
        return total

File: Binary Subarrays With Sum.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        if (goal == 0) {
            int combo = 0;
            int total = 0;
            for (int& num: nums) {
                if (num == 1 && combo > 0) {
                    total += (combo*(combo+1))/2;
                    combo = 0;
                } else if (num == 0) combo++;
            }

            if (combo > 0) total += (combo*(combo+1))/2;
            return total;
        }

        vector<int> arr;
        int combo = 1;
        for (int& num: nums) {
            if (num == 1) {
                arr.push_back(combo);
                combo = 1;
            } else {
                combo++;
            }
        }
        arr.push_back(combo);
        int left = 0;
        int right = goal;
        int total = 0;
        while (right < arr.size()) {
            total += arr.at(right)*arr.at(left);
            right++;
            left++;
        }
        return total;
    }
};